File: app/src/main/kotlin/com/learner/lm/ai/TutorEngine.kt
package com.learner.lm.ai

import com.learner.lm.repository.AiRepository

class TutorEngine(
    private val aiRepository: AiRepository,
    private val promptBuilder: PromptBuilder = PromptBuilder(),
    private val subjectClassifier: SubjectClassifier = SubjectClassifier()
) {

    suspend fun respond(context: TutorContext): TutorResponse {
        val subject = resolveSubject(context)
        val enriched = context.copy(subject = subject)
        val route = ModelRegistry.resolve(
            enriched.appMode,
            enriched.subscriptionTier,
            enriched.freeModelVariant
        )

        val systemPrompt = promptBuilder.buildSystemPrompt(enriched)
        val userPrompt = promptBuilder.buildUserPrompt(enriched)

        val message = aiRepository.generateResponse(
            systemPrompt = systemPrompt,
            userPrompt = userPrompt,
            route = route
        )

        val behaviorMode = enriched.appMode.learningBehavior(enriched.freeModelVariant)
        val sanitized = sanitizeResponse(message, behaviorMode)
        val nextHintLevel = when (behaviorMode) {
            AppMode.TUTOR -> determineNextHintLevel(enriched, sanitized)
            else -> enriched.hintLevel
        }

        return TutorResponse(
            message = sanitized,
            hintLevel = nextHintLevel,
            subject = subject,
            detectedMistake = detectMistakeSeeking(enriched),
            encouragesAttempt = !containsDirectAnswer(sanitized)
        )
    }

    private fun resolveSubject(context: TutorContext): StudySubject {
        if (context.subject is StudySubject.Custom) return context.subject
        val builtin = context.subject as? StudySubject.Builtin ?: return StudySubject.Builtin(Subject.GENERAL)
        if (builtin.subject != Subject.GENERAL) return builtin
        if (context.appMode.learningBehavior(context.freeModelVariant) == AppMode.CODE) {
            return StudySubject.Builtin(Subject.GENERAL)
        }
        val classified = subjectClassifier.classify(
            listOfNotNull(context.studentMessage, context.scannedText).joinToString(" ")
        )
        return StudySubject.Builtin(classified)
    }

    private fun determineNextHintLevel(context: TutorContext, response: String): HintLevel {
        val studentAttempted = context.studentMessage.length > 20 &&
            !isAnswerSeeking(context.studentMessage)
        return when {
            studentAttempted && context.hintLevel == HintLevel.NEAR_SOLUTION ->
                HintLevel.STUDENT_ATTEMPT_REQUIRED
            isAnswerSeeking(context.studentMessage) -> context.hintLevel.next()
            else -> context.hintLevel
        }
    }

    private fun isAnswerSeeking(message: String): Boolean {
        val lower = message.lowercase()
        return lower.contains("what is the answer") ||
            lower.contains("just tell me") ||
            lower.contains("solve this for me") ||
            lower.contains("give me the answer") ||
            lower.contains("write the whole") ||
            lower.contains("build the entire app")
    }

    private fun detectMistakeSeeking(context: TutorContext): String? {
        return when (context.appMode.learningBehavior(context.freeModelVariant)) {
            AppMode.TUTOR -> if (isAnswerSeeking(context.studentMessage)) {
                "It looks like you're asking for a direct answer. Let's work through this step by step instead."
            } else null
            AppMode.CODE -> if (isAnswerSeeking(context.studentMessage)) {
                "I can't build full apps or projects — let's focus on one function or bug at a time."
            } else null
            AppMode.STUDY, AppMode.FREE -> null
        }
    }

    private fun containsDirectAnswer(response: String): Boolean {
        val lower = response.lowercase()
        return lower.contains("the answer is") || lower.contains("the solution is")
    }

    private fun sanitizeResponse(response: String, mode: AppMode): String {
        if (mode != AppMode.TUTOR) return response.trim()
        var sanitized = response
        val forbiddenPatterns = listOf(
            Regex("the answer is\\s+.+", RegexOption.IGNORE_CASE),
            Regex("the solution is\\s+.+", RegexOption.IGNORE_CASE)
        )
        forbiddenPatterns.forEach { pattern ->
            sanitized = sanitized.replace(
                pattern,
                "Let's keep working on this together — what's your next step?"
            )
        }
        return sanitized.trim()
    }
}

File: backend/learner_api/services/tutor_engine.py
from dataclasses import dataclass
import re

from learner_api.schemas import AppMode, HintLevel
from learner_api.services.model_registry import ModelRoute, resolve_model
from learner_api.services.openrouter import OpenRouterClient
from learner_api.services.prompt_builder import PromptBuilder, TutorContext, effective_learning_mode
from learner_api.services.subject_classifier import SubjectClassifier


@dataclass
class TutorResponse:
    message: str
    hint_level: HintLevel
    subject_key: str
    subject_display_name: str
    model_label: str
    detected_mistake: str | None = None
    encourages_attempt: bool = True


class TutorEngine:
    def __init__(
        self,
        openrouter: OpenRouterClient,
        prompt_builder: PromptBuilder | None = None,
        subject_classifier: SubjectClassifier | None = None,
    ) -> None:
        self.openrouter = openrouter
        self.prompt_builder = prompt_builder or PromptBuilder()
        self.subject_classifier = subject_classifier or SubjectClassifier()

    async def respond(self, context: TutorContext, subscription_tier: str) -> TutorResponse:
        subject_key, subject_display, subject_category = self._resolve_subject(context)
        enriched = TutorContext(
            grade_level=context.grade_level,
            subject_key=subject_key,
            subject_display_name=subject_display,
            subject_category=subject_category,
            app_mode=context.app_mode,
            subscription_tier=subscription_tier,
            hint_level=context.hint_level,
            student_message=context.student_message,
            conversation_history=context.conversation_history,
            scanned_text=context.scanned_text,
            free_model_variant=context.free_model_variant,
        )
        route = resolve_model(
            enriched.app_mode,
            subscription_tier,
            enriched.free_model_variant,
        )
        system_prompt = self.prompt_builder.build_system_prompt(enriched)
        user_prompt = self.prompt_builder.build_user_prompt(enriched)

        raw_message = await self.openrouter.chat_completion(
            system_prompt=system_prompt,
            user_prompt=user_prompt,
            route=route,
        )
        sanitized = self._sanitize_response(
            raw_message,
            effective_learning_mode(enriched.app_mode, enriched.free_model_variant),
        )
        learning_mode = effective_learning_mode(enriched.app_mode, enriched.free_model_variant)
        next_hint = (
            self._determine_next_hint_level(enriched, sanitized)
            if learning_mode == AppMode.TUTOR
            else enriched.hint_level
        )

        return TutorResponse(
            message=sanitized,
            hint_level=next_hint,
            subject_key=subject_key,
            subject_display_name=subject_display,
            model_label=route.display_name,
            detected_mistake=self._detect_mistake_seeking(enriched),
            encourages_attempt=not self._contains_direct_answer(sanitized),
        )

    def _resolve_subject(self, context: TutorContext) -> tuple[str, str, str | None]:
        if context.subject_key.startswith("custom:"):
            return context.subject_key, context.subject_display_name, context.subject_category
        builtin = context.subject_key.removeprefix("builtin:")
        learning_mode = effective_learning_mode(context.app_mode, context.free_model_variant)
        if builtin != "GENERAL" or learning_mode == AppMode.CODE:
            return context.subject_key, context.subject_display_name, None
        text = " ".join(filter(None, [context.student_message, context.scanned_text]))
        classified = self.subject_classifier.classify(text)
        return f"builtin:{classified.value}", classified.value.title(), None

    def _determine_next_hint_level(self, context: TutorContext, response: str) -> HintLevel:
        student_attempted = len(context.student_message) > 20 and not self._is_answer_seeking(
            context.student_message
        )
        if student_attempted and context.hint_level == HintLevel.NEAR_SOLUTION:
            return HintLevel.STUDENT_ATTEMPT_REQUIRED
        if self._is_answer_seeking(context.student_message):
            return HintLevel(min(context.hint_level.value + 1, HintLevel.STUDENT_ATTEMPT_REQUIRED.value))
        return context.hint_level

    @staticmethod
    def _is_answer_seeking(message: str) -> bool:
        lower = message.lower()
        phrases = (
            "what is the answer",
            "just tell me",
            "solve this for me",
            "give me the answer",
            "write the whole",
            "build the entire app",
        )
        return any(phrase in lower for phrase in phrases)

    def _detect_mistake_seeking(self, context: TutorContext) -> str | None:
        if not self._is_answer_seeking(context.student_message):
            return None
        learning_mode = effective_learning_mode(context.app_mode, context.free_model_variant)
        if learning_mode == AppMode.TUTOR:
            return "It looks like you're asking for a direct answer. Let's work through this step by step instead."
        if learning_mode == AppMode.CODE:
            return "I can't build full apps or projects — let's focus on one function or bug at a time."
        return None

    @staticmethod
    def _contains_direct_answer(response: str) -> bool:
        lower = response.lower()
        return "the answer is" in lower or "the solution is" in lower

    @staticmethod
    def _sanitize_response(response: str, mode: AppMode) -> str:
        if mode != AppMode.TUTOR:
            return response.strip()
        sanitized = response
        for pattern in (
            re.compile(r"the answer is\s+.+", re.IGNORECASE),
            re.compile(r"the solution is\s+.+", re.IGNORECASE),
        ):
            sanitized = pattern.sub(
                "Let's keep working on this together — what's your next step?",
                sanitized,
            )
        return sanitized.strip()
